src/hotspot/share/gc/serial/serialStringDedup.inline.hpp

```diff
@@ -0,0 +1,49 @@
+/*
+ * Copyright (c) 2021, Alibaba Group Holding Limited. All rights reserved.
+ * DO NOT ALTER OR REMOVE COPYRIGHT NOTICES OR THIS FILE HEADER.
+ *
+ * This code is free software; you can redistribute it and/or modify it
+ * under the terms of the GNU General Public License version 2 only, as
+ * published by the Free Software Foundation.
+ *
+ * This code is distributed in the hope that it will be useful, but WITHOUT
+ * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
+ * FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License
+ * version 2 for more details (a copy is included in the LICENSE file that
+ * accompanied this code).
+ *
+ * You should have received a copy of the GNU General Public License version
+ * 2 along with this work; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301 USA.
+ *
+ * Please contact Oracle, 500 Oracle Parkway, Redwood Shores, CA 94065 USA
+ * or visit www.oracle.com if you need additional information or have any
+ * questions.
+ */
+
+#ifndef SHARE_GC_SERIAL_STRINGDEDUP_INLINE_HPP
+#define SHARE_GC_SERIAL_STRINGDEDUP_INLINE_HPP
+
+#include "gc/serial/serialHeap.hpp"
+#include "gc/serial/serialStringDedup.hpp"
+#include "classfile/javaClasses.inline.hpp"
+#include "oops/oop.inline.hpp"
+
+bool SerialStringDedup::is_candidate_from_mark(oop java_string) {
+  // Candidate if string is being evacuated from young to old but has not
+  // reached the deduplication age threshold, i.e. has not previously been a
+  // candidate during its life in the young generation.
+  return SerialHeap::heap()->young_gen()->is_in_reserved(java_string) &&
```

Comment:
This depends on implicit includes since there's nothing here to ensure the type of `young_gen()` is in scope.  I think it's coming from serialHeap.hpp, but that could be changed in the future to forward-declare the generation class rather than include the associated header.  It might be better to have this function in a .cpp file (as it was in an earlier iteration) to limit header exposure in clients.  This function isn't so performance critical, since before calling it we've already checked that deduplication is enabled and the argument is a String, so being inline isn't as important here as for is_candidate_from_evacuation.

Reply:
Fixed.